crates/ecs/src/lib.rs

```diff
@@ -806,4 +1138,391 @@ mod tests {
+        archetype.add_component_vec::<u64>();
+
+        let a = archetype.borrow_component_vec_mut::<u32>();
+        let b = archetype.borrow_component_vec_mut::<u64>();
+
+        // Drop after both have been borrowed to make sure they both live this long.
+        drop(a);
+        drop(b);
+    }
+
+    #[test]
+    fn removing_entity_swaps_position_of_last_added_entitys_components() {
+        // Arrange
+        let mut archetype = Archetype::default();
+
+        archetype.store_entity(10);
+        archetype.store_entity(20);
+        archetype.store_entity(30);
+
+        archetype.add_component_vec::<u32>();
+        archetype.add_component_vec::<f32>();
+
+        archetype.add_component::<u32>(10, 1);
+        archetype.add_component::<u32>(20, 2);
+        archetype.add_component::<u32>(30, 3);
+
+        archetype.add_component::<f32>(10, 1.0);
+        archetype.add_component::<f32>(20, 2.0);
+        archetype.add_component::<f32>(30, 3.0);
+
+        // Act
+        archetype.remove_entity(10);
+
+        // Assert
+        let component_vec_u32 = archetype.borrow_component_vec::<u32>().unwrap();
+        let component_vec_f32 = archetype.borrow_component_vec::<f32>().unwrap();
+
+        // Removing entity_id 10 should move components of entity_id 30 to index 0.
+        assert_eq!(component_vec_u32.get(0).unwrap().unwrap(), 3);
+        assert_eq!(component_vec_f32.get(0).unwrap().unwrap(), 3.0);
+        // Components of entity_id 20 should stay on index 1.
+        assert_eq!(component_vec_u32.get(1).unwrap().unwrap(), 2);
+        assert_eq!(component_vec_f32.get(1).unwrap().unwrap(), 2.0);
+
+        assert_eq!(component_vec_u32.len(), 2);
+        assert_eq!(component_vec_f32.len(), 2);
+    }
+
+    // Intersection tests:
+    #[test_case(vec![vec![1,2,3]], vec![1,2,3]; "self intersection")]
+    #[test_case(vec![vec![1,2,3], vec![1,2,3]], vec![1,2,3]; "two of the same")]
+    #[test_case(vec![vec![1], vec![2,3], vec![4]], vec![]; "no overlap, no matches")]
+    #[test_case(vec![vec![1,2], vec![2,3], vec![3,4]], vec![]; "some overlap, no matches")]
+    #[test_case(vec![vec![1,2,3,4], vec![2,3], vec![3,4]], vec![3]; "some matches")]
+    #[test_case(vec![Vec::<usize>::new()], vec![]; "empty")]
+    #[test_case(vec![Vec::<usize>::new(), Vec::<usize>::new(), Vec::<usize>::new(), Vec::<usize>::new()], vec![]; "multiple empty")]
+    #[test_case(vec![Vec::<usize>::new(), vec![1,2,3,4]], vec![]; "one empty, one not")]
+    #[test_case(vec![vec![2,1,1,1,1], vec![1,1,1,1,2], vec![1,1,2,1,1]], vec![2,1,1,1,1]; "multiple of the same number")]
+    fn intersection_returns_expected_values(
+        test_vecs: Vec<Vec<usize>>,
+        expected_value: Vec<usize>,
+    ) {
+        // Construct test values, to avoid upsetting Rust and test_case
+        let borrowed_test_vecs: Vec<&Vec<usize>> = test_vecs.iter().collect();
+        let borrowed_expected_value: Vec<&usize> = expected_value.iter().collect();
```

Comment:
nit:
```suggestion
        // Borrow inner values, because `intersection` expects a vector of references.
        let borrowed_test_vecs: Vec<_> = test_vecs.iter().collect();
        let borrowed_expected_value: Vec<_> = expected_value.iter().collect();
```